presto-main/src/main/java/io/prestosql/operator/scalar/ArrayCombinationsFunction.java

```diff
@@ -0,0 +1,129 @@
+import io.prestosql.spi.function.SqlType;
+import io.prestosql.spi.function.TypeParameter;
+import io.prestosql.spi.type.ArrayType;
+import io.prestosql.spi.type.Type;
+
+import java.util.Optional;
+
+import static io.prestosql.spi.StandardErrorCode.INVALID_FUNCTION_ARGUMENT;
+import static io.prestosql.spi.type.StandardTypes.INTEGER;
+import static io.prestosql.util.Failures.checkCondition;
+import static java.lang.Math.multiplyExact;
+import static java.lang.StrictMath.toIntExact;
+import static java.lang.String.format;
+import static java.util.Arrays.setAll;
+
+@ScalarFunction("combinations")
+@Description("Return n-element subsets from array")
+public final class ArrayCombinationsFunction
+{
+    private ArrayCombinationsFunction() {}
+
+    private static final int MAX_COMBINATION_LENGTH = 5;
+    private static final int MAX_RESULT_ELEMENTS = 100_000;
+
+    @TypeParameter("T")
+    @SqlType("array(array(T))")
+    public static Block combinations(
+            @TypeParameter("T") Type elementType,
+            @SqlType("array(T)") Block array,
+            @SqlType(INTEGER) long n)
+    {
+        int arrayLength = array.getPositionCount();
+        int combinationLength = toIntExact(n);
+        checkCondition(combinationLength >= 0, INVALID_FUNCTION_ARGUMENT, "combination length must be positive: %s", combinationLength);
+        checkCondition(combinationLength <= MAX_COMBINATION_LENGTH, INVALID_FUNCTION_ARGUMENT, "combination length must not exceed %s: %s", MAX_COMBINATION_LENGTH, combinationLength);
+
+        ArrayType arrayType = new ArrayType(elementType);
+        if (combinationLength > arrayLength) {
+            return arrayType.createBlockBuilder(new PageBuilderStatus().createBlockBuilderStatus(), 0).build();
+        }
+
+        int combinationCount = combinationCount(arrayLength, combinationLength);
+        checkCondition(combinationCount * (long) combinationLength <= MAX_RESULT_ELEMENTS, INVALID_FUNCTION_ARGUMENT, "combinations exceed max size");
+
+        int[] ids = new int[combinationCount * combinationLength];
+        int idsPosition = 0;
+
+        int[] combination = firstCombination(combinationLength);
+        do {
+            System.arraycopy(combination, 0, ids, idsPosition, combinationLength);
+            idsPosition += combinationLength;
+        }
+        while (nextCombination(combination, arrayLength));
+
+        int[] offsets = new int[combinationCount + 1];
+        setAll(offsets, i -> i * combinationLength);
+
+        return ArrayBlock.fromElementBlock(combinationCount, Optional.empty(), offsets, new DictionaryBlock(array, ids));
+    }
+
+    @VisibleForTesting
+    static int combinationCount(int arrayLength, int combinationLength)
+    {
+        try {
+            int combinations = 1;
+            for (int i = 1; i <= combinationLength; i++) {
+                combinations = multiplyExact(combinations, arrayLength - (combinationLength - i)) / i;
```

Comment:
This is not immediately obvious. Add a comment explaining how it's derived. Something like:

![n_choose_k](https://user-images.githubusercontent.com/14387/57348313-764ed400-710a-11e9-80a2-4e60046a7b5d.png)

BTW, note that you can simplify to:

```
combinations = multiplyExact(combinations, arrayLength - i + 1)) / i;
```

since `combinationLength - i` ranges from  `0 .. (combinationLength - 1)`

Reply:
I added a comment.

I cannot simplify, because the order of multiplications and divisions is significant. If I change order of the arithmetics, I will no longer be able to do this in integer arithmetics.

Reply:
Ok, then that explanation you just have also deserves a comment (or someone will try to improve it in the future and break that unstated assumption)

Reply:
This is why there is a test for that.

Reply:
None of the tests fail if I replace the expression with `multiplyExact(combinations, arrayLength - i + 1) / i`, though.

Reply:
@martint you're right. I must have done some typo when I was testing that previously.
However, I will keep the code as is, because this is way correctness is more obvious -- with recursive approach i calculate n-choose-k value from another n-choose-k value so divisibility is "obvious". With your approach... I would not only need to update the comment (not a big deal) -- I would not know why there is no divisibility problem 😉